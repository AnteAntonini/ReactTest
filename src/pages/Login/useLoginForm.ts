import { useEffect, useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../context/AuthContext";

export const useLoginForm = (message: string) => {
  const navigate = useNavigate();
  const { setIsAuth } = useContext(AuthContext);
  const [errorMessages, setErrorMessages] = useState("");
  const [user, setUser] = useState({
    email: "",
    password: "",
  });

  const registerUsers = [
    {
      email: "user1@gmail.com",
      password: "1234",
    },
    {
      email: "user2@gmail.com",
      password: "1234",
    },
  ];

  const errorMessage =
    "The email address or password is incorrect. Please try again.";

  useEffect(() => {
    console.log(`${message} LoginForm`);
  }, []);

  const handleChange = ({ target }: any) => {
    setUser((prev) => ({ ...prev, [target.name]: target.value }));
  };

  const handleLogin = (e: any) => {
    e.preventDefault();
    e.stopPropagation();

    const userData = registerUsers.find(
      (registerUser) => registerUser.email === user.email
    );

    if (userData) {
      if (userData.password !== user.password) {
        setErrorMessages(errorMessage);
      } else {
        sessionStorage.setItem("isLoggedIn", "true");
        setIsAuth(true);
        navigate("/app");
      }
    } else {
      setErrorMessages(errorMessage);
    }
  };

  return {
    handleLogin,
    handleChange,
    errorMessages,
  };
};
