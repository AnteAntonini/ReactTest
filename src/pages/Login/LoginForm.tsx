import "./loginForm.css";
import { useLoginForm } from "./useLoginForm";

const LoginForm = ({ message }: { message: string }) => {
  const { handleLogin, handleChange, errorMessages } = useLoginForm(message);

  return (
    <form className="login-form" autoComplete="off" onSubmit={handleLogin}>
      <h2 className="header-title">Login</h2>
      <div>
        <div className="error">{errorMessages}</div>
        <div className="row">
          <label>Email</label>
          <input
            autoComplete="off"
            name="email"
            type="email"
            placeholder="Enter your email"
            onChange={handleChange}
          />
        </div>
        <div className="row">
          <label>Password</label>
          <input
            autoComplete="new-password"
            name="password"
            type="password"
            placeholder="Enter your password"
            onChange={handleChange}
          />
        </div>
        <div className="button row">
          <button type="submit">Log In</button>
        </div>
      </div>
    </form>
  );
};

export default LoginForm;
