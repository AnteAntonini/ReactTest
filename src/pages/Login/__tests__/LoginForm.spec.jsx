import { render, fireEvent, waitFor } from "@testing-library/react";
import LoginForm from "../LoginForm";
import { useLoginForm } from "../useLoginForm";

jest.mock("../useLoginForm");

describe("LoginForm Component", () => {
  it("renders login form with initial state", () => {
    // Mock the useLoginForm hook return values
    useLoginForm.mockReturnValue({
      handleLogin: jest.fn(),
      handleChange: jest.fn(),
      errorMessages: "",
    });

    const { getByText, getByPlaceholderText } = render(
      <LoginForm message="Test" />
    );

    // Check if the form elements are present
    expect(getByText("Login")).toBeInTheDocument();
    expect(getByPlaceholderText("Enter your email")).toBeInTheDocument();
    expect(getByPlaceholderText("Enter your password")).toBeInTheDocument();
    expect(getByText("Log In")).toBeInTheDocument();
  });

  it("handles form submission", async () => {
    // Mock the useLoginForm hook return values
    const mockHandleLogin = jest.fn();
    useLoginForm.mockReturnValue({
      handleLogin: mockHandleLogin,
      handleChange: jest.fn(),
      errorMessages: "",
    });

    const { getByText, getByPlaceholderText } = render(
      <LoginForm message="Test" />
    );

    const emailInput = getByPlaceholderText("Enter your email");
    const passwordInput = getByPlaceholderText("Enter your password");
    const submitButton = getByText("Log In");

    // Simulate user input
    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });

    fireEvent.click(submitButton);

    // Wait for the handleLogin function to be called
    await waitFor(() => {
      expect(mockHandleLogin).toHaveBeenCalled();
    });
  });

  it("displays error messages", () => {
    // Mock the useLoginForm hook return values to include an error message
    useLoginForm.mockReturnValue({
      handleLogin: jest.fn(),
      handleChange: jest.fn(),
      errorMessages: "Invalid credentials",
    });

    const { getByText } = render(<LoginForm message="Test" />);

    expect(getByText("Invalid credentials")).toBeInTheDocument();
  });
});

describe("Test snapshot", () => {
  it("Component should match snapshot", () => {
    useLoginForm.mockReturnValue({
      handleLogin: jest.fn(),
      handleChange: jest.fn(),
      errorMessages: "",
    });
    const { asFragment } = render(<LoginForm message="Test" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
