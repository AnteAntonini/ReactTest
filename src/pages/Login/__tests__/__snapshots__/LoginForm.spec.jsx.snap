// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot Component should match snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="login-form"
  >
    <h2
      class="header-title"
    >
      Login
    </h2>
    <div>
      <div
        class="error"
      />
      <div
        class="row"
      >
        <label>
          Email
        </label>
        <input
          autocomplete="off"
          name="email"
          placeholder="Enter your email"
          type="email"
        />
      </div>
      <div
        class="row"
      >
        <label>
          Password
        </label>
        <input
          autocomplete="new-password"
          name="password"
          placeholder="Enter your password"
          type="password"
        />
      </div>
      <div
        class="button row"
      >
        <button
          type="submit"
        >
          Log In
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
